### PR TITLE
Refactor graph construction

### PR DESCRIPTION
**Pull request by chefsalim**
This is a minor refactor of the dependency graph construction that moves it forward toward using the DB to build the package graph.  The PackageGraph implementation now takes in a Package iterator, and the file system walker functionality is moved into a separate module.

Signed-off-by: Salim Alam <salam@chef.io>

giphy 3

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/builder-core/src/file_walker.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    walker: Iter,
+}
+
+impl FileWalker {
+    pub fn new(path: &str) -> Self {
+        FileWalker { walker: WalkDir::new(PathBuf::from(path)).follow_links(false).into_iter() }
+    }
+}
+
+pub fn extract_package(path: &Path) -> Option<depotsrv::Package> {
```
Comment (reset) [inline components/builder-core/src/file_walker.rs]: I'd recommend changing this function signature to use `AsRef<Path>` 

```
pub fn extract_package<T: AsRef<Path>>(path: T) -> Option<depotsrv::Package> {}
```

The reasoning here is that all of the String and Path types in the standard library implement `AsRef<Path>` which make calling this public API function much more ergonomic! I wish I had done this in all the places where I made people pass me a &Path before ;)

File: components/builder-core/src/file_walker.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    let mut archive = PackageArchive::new(PathBuf::from(path));
```
Comment (reset) [inline components/builder-core/src/file_walker.rs]: Coincidentally, anything that implements `AsRef<Path>` also implements `Into<PathBuf>` which is what `PackageArchive::new()` accepts, so you should just be able to pass `path: T` directly to the constructor here if you were to use my advice above

File: components/builder-core/src/file_walker.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+use walkdir::{WalkDir, Iter};
+use hab_core::package::{FromArchive, PackageArchive};
+use protocol::depotsrv;
+
+pub struct FileWalker {
+    walker: Iter,
+}
+
+impl FileWalker {
+    pub fn new(path: &str) -> Self {
```
Comment (reset) [inline components/builder-core/src/file_walker.rs]: I'd recommend `AsRef<Path>` here as well

File: components/builder-core/src/package_graph.rs [hunk 4]
```diff
@@ -93,44 +88,24 @@ impl GraphWalker {
         id
     }
 
-    pub fn build(&mut self) -> (usize, usize) {
+    pub fn build<T>(&mut self, packages: T) -> (usize, usize)
+        where T: Iterator<Item = depotsrv::Package>
```
Comment (reset) [inline components/builder-core/src/package_graph.rs]: Great change here with the use of Iterator

**Comment by reset** [comment]
@chefsalim A few comments inline but they're not blockers - merge whenever you're ready!

**Comment by chefsalim** [comment]
Updated to add AsRef per @reset suggestion.

**Comment by chefsalim** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/199799492).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/199799492).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**